ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -982,15 +943,88 @@ private static void sharedWorkExtendedOptimization(ParseContext pctx, SharedWork
         (Entry<String, TableScanOperator> e) -> e.getValue().getNumChild() == 0);
   }
 
-  /**
-   * Obtain the RS input for a mapjoin operator.
-   */
-  private static ReduceSinkOperator obtainBroadcastInput(MapJoinOperator mapJoinOp) {
+  @VisibleForTesting
+  public void runMapJoinCacheReuseOptimization(
+      ParseContext pctx, SharedWorkOptimizerCache optimizerCache) throws SemanticException {
+    // First we group together all the mapjoin operators whose first broadcast input is the same.
+    final Multimap<Operator<?>, MapJoinOperator> parentToMapJoinOperators = ArrayListMultimap.create();
+    for (Set<Operator<?>> workOperators : optimizerCache.getWorkGroups()) {
+      for (Operator<?> op : workOperators) {
+        if (op instanceof MapJoinOperator) {
+          MapJoinOperator mapJoinOp = (MapJoinOperator) op;
+          // Only allowed for mapjoin operator
+          if (!mapJoinOp.getConf().isBucketMapJoin() && !mapJoinOp.getConf().isDynamicPartitionHashJoin()) {
```

Comment:
please keep the original code formatting

Reply:
Do you mean something like this?
```
            if (!mapJoinOp.getConf().isBucketMapJoin() &&
                !mapJoinOp.getConf().isDynamicPartitionHashJoin()) {
```

Reply:
yes, please